x86: support single or per-domain page tables

tyrantlizard77 opened the pull request:
This PR contains many updates towards getting the x86 implementation to match our plans in https://github.com/zephyrproject-rtos/zephyr/issues/27819. Major updates:

The x86 port used to maintain a separate set of page tables for every user thread. This was ok for 64-bit use-cases where RAM was not much of a concern, but this isn't workable for microcontroller use-cases. This PR implements one of the two page table management schemes described in #27819 
 - By default, page tables are now maintained at the memory domain level, with threads in the same memory domain sharing the same page tables.
 - If `CONFIG_X86_COMMON_PAGE_TABLE` is enabled, there is instead a single set of page tables for all threads/domains that is updated on context switch.
 - For either page table management scheme, x86 allows memory domains to program partitions for any mapped address, not just permanent RAM mappings
 - X86 now implements `struct arch_mem_domain` with page table information if multiple page tables are in use
 - X86 now implements `arch_mem_domain_init()` which creates a fresh set of page tables for the new memory domain, copied from the boot page tables and using the page pool for memory, if we are not using the common page table
 - If multiple page tables are in use, memory mappings are the same across all page tables and all threads. Only permissions with respect to user threads and their memory domain membership is different. Memory access to supervisor mode is the same regardless what thread is scheduled.
 - Since the same page tables could be active on multiple CPUs, implement TLB shootdown logic for SMP.
 - The `z_x86_userspace_enter` implementation now calls back into C code in order to wipe the target stack and update the page table permissions, after switching onto the privilege elevation stack. The US bit for the stack region can only be activated after it has been cleared for security reasons, but before the thread drops privileges.
 - A spinlock now protects the set of active page tables and the slist that links them. All page table modifications are serialized on this lock.
 - We use unused bits in the PTE to backup original page permissions when the mapping was made. This lets us easily restore them when un-applying memory domain partitions.
 - Added build targets to exercise configurations where KPTI is disabled, thanks to @laurenmurphyx64 for the patches
 - Added a test scenario specifically to test memory domain migration, important for SMP if the target thread is running on another CPU.
 - Added qemu_x86_tiny build target, which exercises `CONFIG_X86_COMMON_PAGE_TABLE` with the page pool disabled and a 4MB address space

Fixes: #29175 
Fixes: #29382
Related to: #27819

Commit:
x86: define some unused PTE bits

These are ignored by the CPU and may be used for OS
accounting.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: use unused PTE bits when mapping memory

Page table management for x86 is being revised such that there
will not in many cases be a pristine, master set of page tables.
Instead, when mapping memory, use unused PTE bits to store the
original RW, US, and XD settings when the mapping was made.

This will allow memory domains to alter page tables while still
being able to restore the original mapping permissions.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

laurenmurphyx64 committed:
boards: Add qemu_x86_nokpti target

Adds a build target to run tests on qemu_x86 with KPTI disabled.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 committed:
boards: Add qemu_x86_64_nokpti target

Adds a build target to run tests on qemu_x86_64 with KPTI disabled.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

Commit:
x86: don't reserve room for page tables in stack

These are handled at the memory domain level now.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: userspace: fix tests that don't hold on MMU

MMU threads within the same memory domain have access to
each other's stacks.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: smp: add TLB shootdown logic

This will be needed when we support memory un-mapping, or
the same user mode page tables on multiple CPUs. Neither
are implemented yet.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: move mem_domain->arch first

This sub-struct may have alignment requirements, move it to
the first member.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: move page tables to the memory domain level

- z_x86_userspace_enter() for both 32-bit and 64-bit now
  call into C code to clear the stack buffer and set the
  US bits in the page tables for the memory range.

- Page tables are now associated with memory domains,
  instead of having separate page tables per thread.
  A spinlock protects write access to these page tables,
  and read/write access to the list of active page
  tables.

- arch_mem_domain_init() implemented, allocating and
  copying page tables from the boot page tables.

- struct arch_mem_domain defined for x86. It has
  a page table link and also a list node for iterating
  over them.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mem_protect: add migration test case

We need to make sure that if we migrate a thread to another
memory domain, the migration process doesn't cause the target
thread to explode. This is mostly a concern on SMP systems;
the thread could be running on another CPU at the same time.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Updated PR. Includes test case changes from https://github.com/zephyrproject-rtos/zephyr/pull/29461 as the multiple inits of memory domains in that testcase was breaking a lot of stuff.

Commit:
x86: move page table reservation macros

We don't need this for stacks any more and only use this
for pre-calculating the boot page tables size. Move to C
code, this doesn't need to be in headers anywhere.

Names adjusted for conciseness.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
boards: x86: tune X86_MMU_PAGE_POOL_PAGES

These are set such that we have enough pages in the pool
for typical driver mappings and to instantiate two more
memory domains, which is what our tests require.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: update X86_MMU_PAGE_POOL_PAGES documentation

Help users understand how this should be tuned. Rather than
guessing wildly, set the default to 0. This needs to be tuned
on a per-board, per-application basis anyway.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add logs for tuning page pool size

This will do until we can set up a proper page pool using
all unused ram for paging structures, heaps, and anonymous
mappings.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add qemu_x86_tiny

This build target has all the low-memory options enabled for
memory management: a 4MB address space, 32-bit paging mode,
no KPTI, an empty page pool, and common page tables for
memory domains.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86: add support for common page tables

We provide an option for low-memory systems to use a single set
of page tables for all threads. This is only supported if
KPTI and SMP are disabled. This configuration saves a considerable
amount of RAM, especially if multiple memory domains are used,
at a cost of context switching overhead.

Some caching techniques are used to reduce the amount of context
switch updates; the page tables aren't updated if switching to
a supervisor thread, and the page table configuration of the last
user thread switched in is cached.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86_64: disable shared_mem test until bugs fixed

re-enabling this test was premature, it still crashes very often
due to two known issues affecting x86 64-bit. Filter out for
every 64-bit x86 platform until those bugs are solved, which
will reduce CI failure noise.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: update k_mem_domain_add_partition docs

We no longer require partitions to be within RAM bounds,
MMIO or other mapped memory are fine too.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -646,86 +1054,45 @@ int arch_mem_map(void *virt, uintptr_t phys, size_t size, uint32_t flags)
-	}
+	/* Applied to all page tables as this affects supervisor mode.
+	 * XXX: This never gets reset when the thread exits, which can
+	 * cause problems if the memory is later used for something else.
```
Bug for this: https://github.com/zephyrproject-rtos/zephyr/issues/29499

andrewboie commented on arch/x86/Kconfig:
```diff
@@ -192,12 +192,38 @@ config X86_MMU
+	  Pages will need to drawn from the pool to initialize memory domains.
+	  This does not include the default memory domain if KPTI=n.
+
+	  The specific value used here depends on the size of physical RAM,
```
Tuning this is not intuitive. We can do better. I opened https://github.com/zephyrproject-rtos/zephyr/issues/29526

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -784,393 +1151,472 @@ int arch_buffer_validate(void *addr, size_t size, int write)
+	/* Check if we're doing a migration from a different memory domain
+	 * and have to remove permissions from its old domain.
+	 *
+	 * XXX: The checks we have to do here and in
```
Enhancement link for these `is_migration` hacks: https://github.com/zephyrproject-rtos/zephyr/issues/29601

carlocaione commented on arch/x86/core/x86_mmu.c:
```diff
@@ -529,39 +695,117 @@ static void *page_pool_get(void *context)
+ * @retval -ENOMEM allocation required and no free pages (only if OPTION_ALLOC)
  */
 static int page_map_set(pentry_t *ptables, void *virt, pentry_t entry_val,
-			page_get_func_t get_page, void *ctx)
```
You can now remove the typedef on `page_get_func_t`

andrewboie replied:
should be gone now.

carlocaione commented on arch/x86/core/x86_mmu.c:
```diff
@@ -784,393 +1151,472 @@ int arch_buffer_validate(void *addr, size_t size, int write)
+	pentry_t *old_ptables = (pentry_t *)thread->arch.ptables;
+	bool is_user = (thread->base.user_options & K_USER) != 0;
+	bool is_migration = (old_ptables != NULL) && is_user;
+			     ;
```
spurious `;`

andrewboie replied:
removed.

dcpleung commented on arch/x86/Kconfig:
```diff
@@ -192,12 +192,26 @@ config X86_MMU
+	  This does not include the default memory domain if KPTI=n.
+
+	  The specific value used here depends on the size of physical RAM,
+	  much much additional virtual memory will be mapped at runtime, and
```
"how much"?

chen-png commented:
tested it on acrn x86 platform, it was fine.
